### PR TITLE
Add new abstract base class for methods

### LINKED ISSUE #21772: Create an abstract base class for the mechanics `*Method` classes and create a commit set of attributes & methods
We currently have `KanesMethod`, `LagrangesMethod`, and very soon `JointMethod`. `JointMethod` is unique in that it will use `KanesMethod` or `LagrangesMethod` internally, but, even so, it would be helpful if all three of these methods had some common attributes and methods so that duck typing can work well and there is consistency among the methods.

Some reasons making these more consistent will be helpful:

1. `JointMethod(..., method=KanesMethod|LagrangesMethod|OtherNewMethods)`: `JointMethod` should be able to use a number of other `Method` classes to actually formulate the equations of motion. `JointMethod` simple collects the joints and bodies then passes the correct information to the internal methods. Being able to pass in these and future methods to `JointMethod` would give a clean, simple API.
2. `pydy.system.System` currently consumes `KanesMethod` in its constructor. It should be able to consume any `*Method` object and also the `SymbolicSystem` object (https://github.com/sympy/sympy/blob/master/sympy/physics/mechanics/system.py).
3. The idea for `SymbolicSystem` was/is that each method object could produce the `SymbolicSystem`, e.g. `KanesMethod.export_symbolic_system()`. The symbolic system is supposed to be the most general container that can describe any symbolic multibody system. If that works, then `pydy.system.System()` could consume it instead of (or in addition) to the `*Methods` classes.
4. There are many more `*Methods` classes that could be created, e.g. `HamiltonsMethod`, `TMTMethod`, `NewtonEulerMethod`, etc. So having a common class to build them from is helpful.

Key things that `Method()` should likley have:

- coordinates
- speeds (if not the time derivative of coordinates)
- mass matrix, M of `M*u' = F`
- forcing equations, F of `M*u' = F`
- algebraic constraint equations
- equations for constraint forces (explicit or Lagrange multiplier eqs).
- kinematic different equations (if not q' = u)
- maybe more

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

Partially deals with #21772 

#### Brief description of what is fixed or changed
Added a new class `_Methods` which would serve as an abstract base class for all method(KanesMethod, JointsMethod etc)

#### Release Notes

* physics.mechanics
    * `KanesMethod`  can now take body list and force list in object instantiation.
    * `KanesMethod.kanes_equations` both parameters are optional and can be taken automaticaly if they are passed at object instantiation. 
    * `.bodies` attribute is added to `KanesMethod` which is advised to use instead of `.bodylist` attribute.

### CODE REVIEW

**Comment by sidhu1012** [comment]
@moorepants This one's ready !

**Comment by sidhu1012** [comment]
@moorepants This PR is ready to be merged.

File: sympy/physics/mechanics/kane.py [hunk 1]
```diff
@@ -117,7 +118,7 @@ class KanesMethod:
     def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
             configuration_constraints=None, u_dependent=None,
             velocity_constraints=None, acceleration_constraints=None,
-            u_auxiliary=None):
+            u_auxiliary=None, bodies=None, forcelist=None):
```
Comment (moorepants) [inline sympy/physics/mechanics/kane.py:121]: Docstring about these new kwargs is needed.

File: sympy/physics/mechanics/kane.py [hunk 2]
```diff
@@ -530,6 +531,10 @@ def kanes_equations(self, bodies, loads=None):
             Must be either a non-empty iterable of tuples or None which corresponds
             to a system with no constraints.
         """
+        if bodies is None:
+            bodies = self.bodies
+        if  loads is None and self._forcelist is not None and self._forcelist != []:
```
Comment (moorepants) [inline sympy/physics/mechanics/kane.py:536]: This needs a unit test for these cases.

File: sympy/physics/mechanics/kane.py [hunk 3]
```diff
@@ -530,6 +531,10 @@ def kanes_equations(self, bodies, loads=None):
             Must be either a non-empty iterable of tuples or None which corresponds
             to a system with no constraints.
         """
+        if bodies is None:
```
Comment (moorepants) [inline sympy/physics/mechanics/kane.py:534]: Unit test for this is needed.

File: sympy/physics/mechanics/tests/test_kane.py [hunk 4]
```diff
@@ -21,8 +21,10 @@ def test_one_dof():
     pa = Particle('pa', P, m)
     BL = [pa]
 
-    KM = KanesMethod(N, [q], [u], kd)
-    KM.kanes_equations(BL, FL)
+    KM = KanesMethod(N, [q], [u], kd, bodies=BL, forcelist=FL)
+    KM.kanes_equations()
```
Comment (moorepants) [inline sympy/physics/mechanics/tests/test_kane.py:25]: Both cases should still work, so don't remove the test, only add a new one.

File: sympy/physics/mechanics/tests/test_method.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+from sympy.physics.mechanics.method import _Methods
+from sympy.testing.pytest import raises
+
+def test_method():
+    raises(TypeError, lambda: _Methods())
```
Comment (moorepants) [inline sympy/physics/mechanics/tests/test_method.py:5]: Not sure what the purpose of this is.

Comment (sidhu1012) [reply sympy/physics/mechanics/tests/test_method.py:5]: To make sure `_Method` class is working as desired.

File: sympy/physics/mechanics/tests/test_kane.py (hunk 4 above)
Comment (sidhu1012) [reply sympy/physics/mechanics/tests/test_kane.py:25]: Okay

File: sympy/physics/mechanics/kane.py (hunk 1 above)
Comment (sidhu1012) [reply sympy/physics/mechanics/kane.py:121]: @moorepants  I don't think all other `__init__`  kwargs are documented.

**Comment by sidhu1012** [comment]
@moorepants Review please.

File: sympy/physics/mechanics/kane.py (hunk 1 above)
Comment (moorepants) [reply sympy/physics/mechanics/kane.py:121]: That's unfortunate!

**Comment by moorepants** [comment]
You need to add release notes for this, as it is a new addition.

File: sympy/physics/mechanics/tests/test_method.py (hunk 5 above)
Comment (moorepants) [reply sympy/physics/mechanics/tests/test_method.py:5]: So "as desired" means that if someone tries to use the class by itself a type error should be raised?

**Comment by sidhu1012** [comment]
> You need to add release notes for this, as it is a new addition.

okay

File: sympy/physics/mechanics/tests/test_method.py (hunk 5 above)
Comment (sidhu1012) [reply sympy/physics/mechanics/tests/test_method.py:5]: Yes

**Comment by sidhu1012** [comment]
@moorepants Release notes added.

**Pull request merged**